src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -0,0 +1,430 @@
+                Length = sizeof(bool),
+                Buffer = (byte*)&val
+            };
+            SetParam(QUIC_PARAM_CONN.USE_SEND_BUFFER, buffer);
+        }
+
+        public unsafe void DisableBuffering()
+        {
+            bool val = false;
+            var buffer = new QuicBuffer()
+            {
+                Length = sizeof(bool),
+                Buffer = (byte*)&val
+            };
+            SetParam(QUIC_PARAM_CONN.USE_SEND_BUFFER, buffer);
+        }
+
+        internal override ValueTask ConnectAsync(CancellationToken cancellationToken = default)
+        {
+            SetIdleTimeout(TimeSpan.FromSeconds(120));
+            uint status = _api.ConnectionStartDelegate(
+                _nativeObjPtr,
+                (ushort)_remoteEndPoint.AddressFamily,
+                _remoteEndPoint.Address.ToString(),
+                (ushort)_remoteEndPoint.Port);
+
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            return new ValueTask(_connectTcs.Task);
+        }
+
+        public MsQuicStream StreamOpen(
+            QUIC_STREAM_OPEN_FLAG flags)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            IntPtr streamPtr = IntPtr.Zero;
+            uint status = _api.StreamOpenDelegate(
+                _nativeObjPtr,
+                (uint)flags,
+                MsQuicStream.NativeCallbackHandler,
+                IntPtr.Zero,
+                out streamPtr);
+
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            MsQuicStream stream = new MsQuicStream(_api, this, flags, streamPtr, inbound: false);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return stream;
+        }
+
+        public void SetCallbackHandler()
+        {
+            _handle = GCHandle.Alloc(this);
+            _connectionDelegate = new ConnectionCallbackDelegate(NativeCallbackHandler);
+            _api.SetCallbackHandlerDelegate(
+                _nativeObjPtr,
+                _connectionDelegate,
+                GCHandle.ToIntPtr(_handle));
+        }
+
+        public Task ShutdownAsync(
+            QUIC_CONNECTION_SHUTDOWN_FLAG Flags,
+            ushort ErrorCode)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            uint status = _api.ConnectionShutdownDelegate(
+                _nativeObjPtr,
+                (uint)Flags,
+                ErrorCode);
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return _shutdownTcs.Task;
+        }
+
+        internal static uint NativeCallbackHandler(
+            IntPtr connection,
+            IntPtr context,
+            ref ConnectionEvent connectionEventStruct)
+        {
+            var handle = GCHandle.FromIntPtr(context);
+            var quicConnection = (MsQuicConnection)handle.Target;
+            // TODO quicConnection can be null
```

Comment:
How can `quicConnection` be `null`? `connection` cannot be `null`.

Reply:
We _may_ need to protect against the against the MsQuicConnection being disposed (not the native connection object). If the MsQuicConnection is disposed and we get a callback, the MsQuicConnection will be null. I had a bug before which was causing this though, need to confirm if this happens in the abortive shutdown case.

Reply:
I'd recommend **not** freeing the MsQuicConnection until you get a SHUTDOWN_COMPLETE event.

Reply:
Sounds like Dispose will need to wait for the SHUTDOWN_COMPLETE even in the abortive case?